docs/SPEC.md

```diff
@@ -1424,6 +1430,87 @@ A query specification defines what data and operations to perform.
+
+An edition is separate from a version.
+A version of Flux represents a single point in the commit history of the Flux source code.
+A user can only use a single version of Flux for a given script.
+Editions represent a set of features that are enabled for a given Flux version.
+So long as the Flux version supports all the features of the edition it can be enabled.
+A user can upgrade to a newer Flux version without being required to upgrade to the newest edition of Flux.
+
+New Flux features that require a breaking change to the syntax or semantics of Flux must always be part of a new edition.
+This means that a script can regularly update to the newest Flux version without risk of breaking because any breaking changes are explicitly opt-in.
+With editions the Flux community gets both a pattern where users can always be running the latest version of Flux and the ability to introduce new useful but otherwise breaking changes to Flux.
+
+We anticipate that there will be at most one new editions of Flux a year. A slow cadence of new editions means users have ample time to migrate to a new edition if desired.
+Even being a few years behind on editions should only mean a few migration steps in order to have access to features introduced in the newest edition.
+
+### First Edition
+
+The first edition of Flux is 2022.1.
+This first edition is the only minimum required edition of Flux.
+
+### Future Editions
+
+Editions will be named after the year in which they are created, with an added sequence number if more than one edition needs to be created in a single year.
+For example the first edition is `2022.1` because it is created in the year `2022` and is the first edition created in that year.
+
+### Editions are optional
+
+A Flux script may specify directly the edition it requires to function.
+Additionally the Flux runtime will allow for the edition to be specified out of band of the script, thus allowing for deployments of Flux to have control over the edition.
+
+#### Edition attribute
+
+The edition of the current script is specified as an `edition` attribute on the package with a single parameter, the name of the edition.
+It is an error for multiple files within a package or module to specify differing editions.
+
+Examples:
+
+Specify the edition with an explicit package clause
+
+    @edition("2022.2")
+    package math
+
+    add = (x,y) => x + y
+
+Specify the edition without the need for a package clause
+
+    @edition("2022.2")
+
+    import "array"
+
+    array.from(rows: [{_value: 1.0}])
+
+
+### Flux Editions and Modules
+
+Each Flux module may specify its own edition, therefore a Flux script on an earlier edition may import and consume a module that uses a newer edition of Flux.
+Naturally if a module exposes a new edition feature via its API, consumers of that module will be required to use at least that edition in order to directly consume the module.
```

Comment:
I am trying to understand backwards/forwards compatibility with respect to editions.

Suppose in 2023.1 we have polymorphic labels.

If a module exposes polymorphic labels in its API, then `main` will need to be at least on `2023.1`. This makes sense to me.

What if we have a module that is on `2022.1` and uses `keep()` in its implementation (i.e., does not use polymorphic labels), and `main` declares itself to be on `2023.1`? Is the runtime able to use the older version of `keep()` on the module but still use polymorphic labels in `main`?

Reply:
>  Is the runtime able to use the older version of keep() on the module but still use polymorphic labels in main?

Yes, because the only difference with `keep` should be syntactic/semantic. The behavior of the old and new versions of keep is the same. The invariant of editions is that its surface layer i.e. syntactic changes only. The runtime is tied to the version and only a single version is used for any given script. 

Maybe you can expand with an example what might break in this scenario?

Reply:
Ah, the difference being syntactic is the piece I was missing. As long as the underlying representation stays the same, like string arguments being translated to labels or vice versa, that would be fine.